routers/private/hook.go

```diff
@@ -155,6 +156,42 @@ func HookPreReceive(ctx *gitea_context.PrivateContext) {
 			private.GitQuarantinePath+"="+opts.GitQuarantinePath)
 	}
 
+	if git.SupportProcReceive {
+		// if need check write permission
+		for _, refFullName := range opts.RefFullNames {
+			if !strings.HasPrefix(refFullName, git.PullRequestPrefix) {
+				pusher, err := models.GetUserByID(opts.UserID)
+				if err != nil {
+					log.Error("models.GetUserByID：%v", err)
+					ctx.Error(http.StatusInternalServerError, "")
+					return
+				}
+
+				perm, err := models.GetUserRepoPermission(repo, pusher)
+				if err != nil {
+					log.Error("models.GetUserRepoPermission:%v", err)
+					ctx.Error(http.StatusInternalServerError, "")
+					return
+				}
+
+				if !perm.CanWrite(models.UnitTypeCode) {
```

Comment:
Maybe we should check if he has write permission to pull request but not to the code directly.

Reply:
This check is delayed permission check when user want to update other ref (banch, tag ...). so write permission to code is enough.